language/extensions/move-table-extension/sources/Table.move

```diff
@@ -43,12 +43,22 @@ module Extensions::Table {
         &borrow_box<K, V, Box<V>>(table, key).val
     }
 
+    /// Same as `borrow` but pass in a copyable key instead.
+    public fun borrow_by_key<K: copy, V>(table: &Table<K, V>, key: K): &V {
```

Comment:
I'm not really a fan of this name..

Reply:
me, either.

Reply:
I think we should switch the entire API to use copyable keys everywhere. If we leave it like suggested here, it will create a bunch of confusion, AFAICT this API is rarely used for now, so we should be able to do this.

The original functionality of being able to use arbitrary non-copyable data as keys would then need to be achieved by `borrow(table, sha256(&non_copyable_key))`.

Reply:
> I think we should switch the entire API to use copyable keys everywhere. If we leave it like suggested here, it will create a bunch of confusion, AFAICT this API is rarely used for now, so we should be able to do this.
> 
> The original functionality of being able to use arbitrary non-copyable data as keys would then need to be achieved by `borrow(table, sha256(&non_copyable_key))`.

I totally support adding `copy` to the key type.

Reply:
> > I think we should switch the entire API to use copyable keys everywhere. If we leave it like suggested here, it will create a bunch of confusion, AFAICT this API is rarely used for now, so we should be able to do this.
> > The original functionality of being able to use arbitrary non-copyable data as keys would then need to be achieved by `borrow(table, sha256(&non_copyable_key))`.
> 
> I totally support adding `copy` to the key type.

It's sad that we have to do this because of Move borrow semantics restriction. From a language design viewpoint, we are hitting a problem of lacking compositionality. Basically, looking at `borrow(table, sha256(&non_copyable_key))`, we are not able to define a new function `borrow'(table, &k) == borrow(table, sha256(k))`. This is a serious language restriction which we should address, at least for library builders (this kind of developers could use attributes, no general lifetime annotations needed). However, we won't be able to address this in the same time frame as we need tables ready. 

@sblackshear @tnowacki @davidiw

Reply:
@tnowacki I thought you were thinking about implementing lifetime in `move` long time ago?

Reply:
Actually, this is a great point, I'd be curious from @msmouse 's perspective on the cost associated with the following:
* copyable
* reference
* reference + bcs + sha256 (would this satisfy the borrow checker requirements?)

Reply:
we probably still want to have the by reference API after borrow checker upgrades?

Reply:
We don't know when, if, and how the borrow checker will be upgraded. If it is we can introduce the reference API again. In the meantime, it just creates confusion for Move users, as it is not fully usable.

Reply:
Let me also add that the decision to use `&K` was not unanimous. Some folks preferred the `K has copy` approach from the beginning.

Reply:
I like the route where we use `K has copy` in all the APIs, because as a beginner Mover those APIs will hint more explicitly that the size of the key matters. And smaller keys doesn't need to go through sha256 to waste CPU and be inflated in size (@davidiw did you imply a solution where we always sha256 the keys?)